File: Gas Station.cpp
// Leetcode Problem 134: Gas Station
// C++ CODE
#include <vector>
#include <numeric>

class Solution {
public:
    int canCompleteCircuit(std::vector<int>& gas, std::vector<int>& cost) {
        int totalTank = 0;
        int currentTank = 0;
        int startingStation = 0;

        for (int i = 0; i < gas.size(); ++i) {
            int netGain = gas[i] - cost[i];
            totalTank += netGain;
            currentTank += netGain;

            // If fuel becomes negative, reset starting candidate to i + 1
            if (currentTank < 0) {
                startingStation = i + 1;
                currentTank = 0;
            }
        }

        // If overall gas is insufficient to cover total cost, complete circuit is impossible
        return (totalTank >= 0) ? startingStation : -1;
    }
};

File: Gas Station.py
# Leetcode Problem 134: Gas Station
# PYTHON CODE
from typing import List


class Solution:

    def canCompleteCircuit(self, gas: List[int], cost: List[int]) -> int:
        # If total gas is less than total cost, complete circuit is impossible
        if sum(gas) < sum(cost):
            return -1

        total_tank = 0
        current_tank = 0
        starting_station = 0

        for i in range(len(gas)):
            net_gain = gas[i] - cost[i]
            total_tank += net_gain
            current_tank += net_gain

            # If current tank falls below 0, reset starting position to i + 1
            if current_tank < 0:
                starting_station = i + 1
                current_tank = 0

        return starting_station